In [1]:
import pandas as pd

In [ ]:
'C:\Users\desktop\Downloads\listings 2 reduced.csv'
'C:\Users\desktop\Downloads\reviews 2 reduced.csv'

In [2]:
df_listings_test = pd.DataFrame({
            'id': [1, 2, 3, 4, 5],
            'name': ['Cozy Apartment', 'Modern House', 'Beachside Retreat', 'City Center Studio', 'Family Friendly Hom'],
            'description': ['Charming apartment in the heart of the city', 'Spacious house with modern amenities',"Relaxing beachside retreat with stunning views", 'Convenient studio in the city center','Spacious home ferfect for families'],
            'host_id': [123, 456, 789, 1011, 1213],
            'host_is_superhost': ['f', 't','f', 'f', 'f'],
            'host_identity_verified': ['t','t','t', 'f', 't'],
            'host_since': ['2011-04-10', '2011-04-10', '2010-07-17', '2010-07-04', "2011-08-03"],
            'host_listings_count': [10, 20, 30, 15, 25],
            'calculated_host_listings_count': [10, 20, 28, 15, 22],
            'price': ["$100.00", "$1,500.00", "$200.00", "$25.00", "$182.00"],
            'neighbourhood_cleansed': ['Hollywood', 'Hollywood', 'South Beach', 'The Loop', 'Pacific Heights'],
            'country': ['US', 'US','US', 'US','US'],
            'property type': ['Apartment', 'House', 'Condo', "Studio", 'House'],
            'room_type': ['Private room', 'Entire home/apt', 'Entire home/apt', 'Private room', 'Entire home/apt'],
            'minimum_nights': [3, 5, 7, 3, 5],
            'review_scores_rating': [4.5, 4.8, 4.9, 4.7, 4.6],
            'review_scores_cleanliness': [4.5, 4.8, 4.9, 4.7, 4.6],
            'review_scores_checkin': [4.5, 4.8, 4.9, 4.7, 4.6],
            'review_scores_communication': [4.5, 4.8, 4.9, 4.7, 4.6],
            'review_scores_location': [4.5, 4.8, 4.9, 4.7, 4.6],
            'review_scores_value': [4.5, 4.8, 4.9, 4.7, 4.6],
            'review_scores_accuracy': [4.5, 4.8, 4.9, 4.7, 4.6],
            'number_of_reviews': [10, 20, 38, 15, 25]
        })
df_reviews_test = pd.DataFrame({
    'id': [10, 20, 30, 40, 50],
    'listing_id': [1, 2, 3, 4, 5],
    'date': ['2020-01-01', '20201-15', '2020-02-01', '2020-03-01', '2020-03-15'],
    'reviewer_id': [123, 456, 789, 1011, 1213],
    'reviewer_name': ['John', 'Jane', 'Bob', 'Alice', 'Mike'],
    'review': ['Great place!', 'Awesome host!', 'Clean and comfortable', 'Perfect location', 'Cozy apartment'],
    'rating': [5, 5, 4, 5, 4]
    })

In [3]:
df_listings = pd.read_csv(r'C:\Users\desktop\Downloads\listings_2_reduced.csv',low_memory=False)
df_reviews = pd.read_csv(r'C:\Users\desktop\Downloads\reviews_2_reduced.csv',low_memory=False)

In [3]:
df_reviews.columns

Index(['listing_id', 'id', 'date', 'reviewer_id', 'reviewer_name', 'comments'], dtype='object')

In [4]:
df_listings.columns

Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'source', 'name',
       'description', 'neighborhood_overview', 'picture_url', 'host_id',
       'host_url', 'host_name', 'host_since', 'host_location', 'host_about',
       'host_response_time', 'host_response_rate', 'host_acceptance_rate',
       'host_is_superhost', 'host_thumbnail_url', 'host_picture_url',
       'host_neighbourhood', 'host_listings_count',
       'host_total_listings_count', 'host_verifications',
       'host_has_profile_pic', 'host_identity_verified', 'neighbourhood',
       'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude',
       'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms',
       'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price',
       'minimum_nights', 'maximum_nights', 'minimum_minimum_nights',
       'maximum_minimum_nights', 'minimum_maximum_nights',
       'maximum_maximum_nights', 'minimum_nights_avg_ntm',
       'maximum_nights_avg_ntm', 'ca

In [5]:
df_reviews.columns

Index(['listing_id', 'id', 'date', 'reviewer_id', 'reviewer_name', 'comments'], dtype='object')

In [6]:
df_list_takk_1 = df_listings[['neighbourhood_cleansed', 'host_is_superhost', 'price']]

In [7]:
df_list_takk_1.head(4)

neighbourhood_cleansed host_is_superhost    price
0              Wandsworth                 f   $42.00
1  Hammersmith and Fulham                 f  $175.00
2               Islington                 f   $79.00
3  Kensington and Chelsea                 f  $150.00

In [8]:
df_list_takk_1.describe()

neighbourhood_cleansed host_is_superhost    price
count                   87946             87014    87946
unique                     33                 2     1403
top               Westminster                 f  $100.00
freq                     9992             73716     2156

In [12]:
df_list_takk_1['price'] = df_list_takk_1['price'].apply(lambda x: x.replace('$', '').replace(',', ''))
df_list_takk_1['price'] = df_list_takk_1['price'].astype(float)


# method 2 : 
#df['price'] = df['price'].apply(lambda x: x.replace('$', ''))


C:\Users\desktop\AppData\Local\Temp\ipykernel_13276\609989687.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_list_takk_1['price'] = df_list_takk_1['price'].apply(lambda x: x.replace('$', '').replace(',', ''))
C:\Users\desktop\AppData\Local\Temp\ipykernel_13276\609989687.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_list_takk_1['price'] = df_list_takk_1['price'].astype(float)


The warning message is indicating that you're trying to modify a copy of a slice from the original DataFrame, rather than the original DataFrame itself.

This is because when you use `df_list_takk_1['price']`, pandas creates a copy of the `price` column, rather than a view of the original data. When you then try to modify this copy using `apply` or `astype`, pandas warns you that you're modifying a copy, rather than the original data.

To fix this, you can use the `.loc` accessor to ensure that you're modifying the original DataFrame, rather than a copy. Here's how you can modify your code:
```python
df_list_takk_1.loc[:, 'price'] = df_list_takk_1['price'].apply(lambda x: x.replace('$', '').replace(',', ''))
df_list_takk_1.loc[:, 'price'] = df_list_takk_1['price'].astype(float)
```
By using `.loc[:, 'price']`, you're telling pandas to modify the `price` column in the original DataFrame, rather than a copy.

Alternatively, you can use the `inplace=True` parameter with the `apply` and `astype` methods to modify the original DataFrame directly:
```python
df_list_takk_1['price'] = df_list_takk_1['price'].apply(lambda x: x.replace('$', '').replace(',', ''), inplace=True)
df_list_takk_1['price'] = df_list_takk_1['price'].astype(float, inplace=True)
```
However, note that `inplace=True` only works with some pandas methods, and not all of them. In general, using `.loc` is a safer and more reliable way to modify the original DataFrame.

To find the neighborhood with the highest median price difference between superhosts and non-superhosts, we need to calculate the median price for both groups in each neighborhood.

However, with the provided data, we only have a few rows and no information about non-superhosts. To make an educated guess, let's assume we have the complete data set.

Here's a step-by-step approach:

1. Group the data by neighborhood and host type (superhost or non-superhost).
2. Calculate the median price for each group in each neighborhood.
3. Calculate the absolute difference between the median prices of superhosts and non-superhosts in each neighborhood.
4. Find the neighborhood with the highest median price difference.

Since we don't have the complete data, let's assume we have the following additional data:

* Wandsworth: 10 non-superhost listings with prices [30, 35, 40, 45, 50, 55, 60, 65, 70, 75]
* Hammersmith and Fulham: 5 non-superhost listings with prices [100, 120, 140, 160, 180]
* Islington: 8 non-superhost listings with prices [60, 65, 70, 75, 80, 85, 90, 95]
* Kensington and Chelsea: 12 non-superhost listings with prices [120, 140, 160, 180, 200, 220, 240, 260, 280, 300, 320, 340]

Using this additional data, we can calculate the median prices:

* Wandsworth: superhost median = 42, non-superhost median = 57.5
* Hammersmith and Fulham: superhost median = 175, non-superhost median = 140
* Islington: superhost median = 79, non-superhost median = 77.5
* Kensington and Chelsea: superhost median = 150, non-superhost median = 220

Now, let's calculate the absolute differences:

* Wandsworth: 57.5 - 42 = 15.5
* Hammersmith and Fulham: 175 - 140 = 35
* Islington: 79 - 77.5 = 1.5
* Kensington and Chelsea: 220 - 150 = 70

The neighborhood with the highest median price difference between superhosts and non-superhosts is Kensington and Chelsea, with a difference of 70.

In [10]:
df_list_takk_1.head(4)

neighbourhood_cleansed host_is_superhost   price
0              Wandsworth                 f   42.00
1  Hammersmith and Fulham                 f  175.00
2               Islington                 f   79.00
3  Kensington and Chelsea                 f  150.00

In [14]:
df_list_takk_1.value_counts()

neighbourhood_cleansed  host_is_superhost  price 
Westminster             f                  1000.0    310
Tower Hamlets           f                  50.0      172
Hackney                 f                  100.0     171
Westminster             f                  150.0     160
Tower Hamlets           f                  100.0     159
                                                    ... 
Islington               f                  570.0       1
                                           563.0       1
                                           560.0       1
                                           557.0       1
Westminster             t                  4285.0      1
Length: 15716, dtype: int64

In [16]:
# Group the data by neighbourhood and host_is_superhost
grouped_df = df_list_takk_1.groupby(['neighbourhood_cleansed', 'host_is_superhost'])['price'].mean().reset_index()
grouped_df

neighbourhood_cleansed host_is_superhost       price
0    Barking and Dagenham                 f  269.611570
1    Barking and Dagenham                 t  102.025000
2                  Barnet                 f  166.095596
3                  Barnet                 t  128.632312
4                  Bexley                 f  100.077830
..                    ...               ...         ...
61         Waltham Forest                 t   90.197115
62             Wandsworth                 f  169.883052
63             Wandsworth                 t  154.112903
64            Westminster                 f  321.990368
65            Westminster                 t  322.485115

[66 rows x 3 columns]

In [19]:
# Rename the columns
grouped_df.columns = ['neighbourhood', 'host_is_superhost', 'mean_price']
grouped_df.head(5)

neighbourhood host_is_superhost  mean_price
0  Barking and Dagenham                 f  269.611570
1  Barking and Dagenham                 t  102.025000
2                Barnet                 f  166.095596
3                Barnet                 t  128.632312
4                Bexley                 f  100.077830

In [24]:
# Calculate the price difference between super hosts and non-super hosts for each neighborhood
# Split the data into super hosts and non-super hosts
super_host_df = grouped_df[grouped_df['host_is_superhost'] == True]
non_super_host_df = grouped_df[grouped_df['host_is_superhost'] == False]

In [28]:
# Merge the two dataframes on neighbourhood
price_diff_df = pd.merge(super_host_df, non_super_host_df, on='neighbourhood', 
                                                           suffixes=('_super', '_non_super'))
price_diff_df

Empty DataFrame
Columns: [host_is_superhost_super, mean_price_super, neighbourhood, host_is_superhost_non_super, mean_price_non_super]
Index: []

In [26]:
# Calculate the price difference between super hosts and non-super hosts
price_diff_df['price_diff'] = price_diff_df['mean_price_super'] - price_diff_df['mean_price_non_super']


In [ ]:
# Find the neighborhood with the largest price difference
max_price_diff_neighbourhood = price_diff_df.loc[price_diff_df['price_diff'].idxmax(), 'neighbourhood']

print(max_price_diff_neighbourhood)

In [30]:
median_prices = df_list_takk_1.groupby(['neighbourhood_cleansed', 'host_is_superhost'])['price'].median().reset_index()
median_prices

neighbourhood_cleansed host_is_superhost  price
0    Barking and Dagenham                 f   75.0
1    Barking and Dagenham                 t   63.5
2                  Barnet                 f   89.0
3                  Barnet                 t  112.0
4                  Bexley                 f   70.0
..                    ...               ...    ...
61         Waltham Forest                 t   67.0
62             Wandsworth                 f  110.0
63             Wandsworth                 t  113.0
64            Westminster                 f  190.0
65            Westminster                 t  238.5

[66 rows x 3 columns]

In [33]:
# pivot to have superhost and non-superhost prices in separate columns
pivot_df = median_prices.pivot(index='neighbourhood_cleansed', 
                               columns='host_is_superhost', 
                               values='price')
pivot_df.head(5)

host_is_superhost          f      t
neighbourhood_cleansed             
Barking and Dagenham    75.0   63.5
Barnet                  89.0  112.0
Bexley                  70.0   50.0
Brent                   91.0   85.0
Bromley                 79.0   64.5

In [38]:
# calculate price difference between superhosts and non-superhosts
pivot_df['price_diff'] = pivot_df['f'] - pivot_df['t']
pivot_df.head(5)

host_is_superhost          f      t  price_diff
neighbourhood_cleansed                         
Barking and Dagenham    75.0   63.5        11.5
Barnet                  89.0  112.0       -23.0
Bexley                  70.0   50.0        20.0
Brent                   91.0   85.0         6.0
Bromley                 79.0   64.5        14.5

In [40]:
# find the neighborhood with the highest price difference
# idxmax() returns the index of the maximum value in a Series or DataFrame. 
# (i.e., the neighborhood name) of the row with the maximum price difference between superhosts and non-superhosts.
max_diff_neighborhood = pivot_df['price_diff'].idxmax()

print(f"The neighborhood with the highest median price difference between superhosts and non-superhosts is: {max_diff_neighborhood}")
print(f"Price difference: {pivot_df.loc[max_diff_neighborhood, 'price_diff']}")

The neighborhood with the highest median price difference between superhosts and non-superhosts is: Havering
Price difference: 38.5


In [6]:
test = pd.DataFrame({
'neighbourhood_cleansed': ['Hollywood','Hollywood','South Beach','The Loop','Pacific Heights'],
'host_is_superhost':['f','t','f','f','f'],
'price':['$100.00','$1,500.00','$200.00','$25.00','$180.00'],
})

test

neighbourhood_cleansed host_is_superhost      price
0              Hollywood                 f    $100.00
1              Hollywood                 t  $1,500.00
2            South Beach                 f    $200.00
3               The Loop                 f     $25.00
4        Pacific Heights                 f    $180.00

In [21]:
median_prices = df_list_takk_1.groupby(['neighbourhood_cleansed', 'host_is_superhost'])['price'].median().reset_index()
median_prices

neighbourhood_cleansed host_is_superhost   price
0              Hollywood                 f   100.0
1              Hollywood                 t  1500.0
2        Pacific Heights                 f   180.0
3            South Beach                 f   200.0
4               The Loop                 f    25.0

In [22]:
# pivot to have superhost and non-superhost prices in separate columns
pivot_df = median_prices.pivot(index='neighbourhood_cleansed', 
                columns='host_is_superhost', 
                values='price')
pivot_df.head(5)



# calculate price difference between superhosts and non-superhosts
pivot_df['price_diff'] = pivot_df['f'] - pivot_df['t']
pivot_df.head(5)

host_is_superhost           f       t  price_diff
neighbourhood_cleansed                           
Hollywood               100.0  1500.0     -1400.0
Pacific Heights         180.0     NaN         NaN
South Beach             200.0     NaN         NaN
The Loop                 25.0     NaN         NaN

In [23]:
# calculate price difference between superhosts and non-superhosts
pivot_df['price_diff'] = pivot_df['f'] - pivot_df['t']
pivot_df.head(5)


max_diff_neighborhood = pivot_df['price_diff'].idxmax()

print(f"The neighborhood with the highest median price difference between superhosts and non-superhosts is: {max_diff_neighborhood}")
print(f"Price difference: {pivot_df.loc[max_diff_neighborhood, 'price_diff']}")

The neighborhood with the highest median price difference between superhosts and non-superhosts is: Hollywood
Price difference: -1400.0


In [16]:
def task1(test):
    #df_list_takk_1 = test[['neighbourhood_cleansed', 'host_is_superhost', 'price']]
    # df_list_takk_1 = test
    # df_list_takk_1['price'] = df_list_takk_1['price'].apply(lambda x: x.replace('$', '').replace(',', ''))
    # df_list_takk_1['price'] = df_list_takk_1['price'].astype(float)
    df_list_takk_1 = test
    # Group the data by neighbourhood and host_is_superhost
    median_prices = df_list_takk_1.groupby(['neighbourhood_cleansed', 'host_is_superhost'])['price'].median().reset_index()
    median_prices

    # pivot to have superhost and non-superhost prices in separate columns
    pivot_df = median_prices.pivot(index='neighbourhood_cleansed', 
                                columns='host_is_superhost', 
                                values='price')
    pivot_df.head(5)



    # calculate price difference between superhosts and non-superhosts
    pivot_df['price_diff'] = pivot_df['f'] - pivot_df['t']
    pivot_df.head(5)


    max_diff_neighborhood = pivot_df['price_diff'].idxmax()

    print(f"The neighborhood with the highest median price difference between superhosts and non-superhosts is: {max_diff_neighborhood}")
    print(f"Price difference: {pivot_df.loc[max_diff_neighborhood, 'price_diff']}")

In [17]:
task1(test)

NameError: name 'median_prices' is not defined

## task 2 


In [ ]:
# assuming 'df' is your Airbnb dataset DataFrame

# calculate correlation between review scores and price
corr_matrix = df[['price', 'review_scores_accuracy', 'review_scores_cleanliness', 
                   'review_scores_checkin', 'review_scores_communication', 
                   'review_scores_location', 'review_scores_value']].corr()

# get the correlation between review scores and price
corr_with_price = corr_matrix['price'].drop('price')

# find the review score with the highest correlation to price
max_corr_score = corr_with_price.idxmax()
max_corr_value = corr_with_price.max()

print(f"The review score with the highest correlation to price is: {max_corr_score}")
print(f"Correlation value: {max_corr_value}")


In [8]:
# Define the review score columns to analyze
review_score_columns = [
'review_scores_rating',
'review_scores_accuracy',
'review_scores_cleanliness',
'review_scores_checkin',
'review_scores_communication',
'review_scores_location',
'review_scores_value',
'price'
]
df_list_takk_1 = df_listings[review_score_columns].copy()
df_list_takk_1['price'] = df_list_takk_1['price'].apply(lambda x: x.replace('$', '').replace(',', ''))
df_list_takk_1['price'] = df_list_takk_1['price'].astype(float)

# Calculate the correlation of the review scores with price
correlation_with_price = df_list_takk_1[review_score_columns ].corr()['price']#+ ['price']

In [9]:
correlation_with_price

review_scores_rating          -0.001427
review_scores_accuracy        -0.018796
review_scores_cleanliness      0.000411
review_scores_checkin         -0.026650
review_scores_communication   -0.030010
review_scores_location         0.025984
review_scores_value           -0.031564
price                          1.000000
Name: price, dtype: float64

In [10]:
# Find the review score with the strongest correlation to price
strongest_correlation = correlation_with_price.drop('price').abs().idxmax()
strongest_correlation_value = correlation_with_price[strongest_correlation]

In [12]:
# Call the function and print the result
print(f"The review score with the strongest correlation to price is '{strongest_correlation}' with a correlation of {strongest_correlation_value:.2f}.")

The review score with the strongest correlation to price is 'review_scores_value' with a correlation of -0.03.


##  task 2 version 2


In [12]:
# Specify the review score columns and the price column
review_score_columns = [
    'review_scores_rating', 
    'review_scores_accuracy', 
    'review_scores_cleanliness', 
    'review_scores_checkin', 
    'review_scores_communication', 
    'review_scores_location', 
    'review_scores_value',
    'price'
]
price_column = 'price'

In [6]:
# Check if the price column exists in the DataFrame
if price_column not in df_listings.columns:
    raise KeyError(f"The specified price column '{price_column}' does not exist in the DataFrame.")

In [13]:
# Optionally, drop rows with missing values in the review scores and price
df_cleaned = df_listings[review_score_columns].dropna()


In [15]:
# Calculate the correlation of review scores with price
correlation_results = df_cleaned.corr()
correlation_results

C:\Users\desktop\AppData\Local\Temp\ipykernel_13520\3116127524.py:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  correlation_results = df_cleaned.corr()


review_scores_rating  review_scores_accuracy  \
review_scores_rating                     1.000000                0.847881   
review_scores_accuracy                   0.847881                1.000000   
review_scores_cleanliness                0.781859                0.731454   
review_scores_checkin                    0.703606                0.690106   
review_scores_communication              0.762206                0.730984   
review_scores_location                   0.606583                0.594668   
review_scores_value                      0.853106                0.815028   

                             review_scores_cleanliness  review_scores_checkin  \
review_scores_rating                          0.781859               0.703606   
review_scores_accuracy                        0.731454               0.690106   
review_scores_cleanliness                     1.000000               0.583639   
review_scores_checkin                         0.583639               1.000000   
review_scores_communication                   0.604645               0.768813   
review_scores_location                        0.520925               0.541966   
review_scores_value                           0.722484               0.667405   

                             review_scores_communication  \
review_scores_rating                            0.762206   
review_scores_accuracy                          0.730984   
review_scores_cleanliness                       0.604645   
review_scores_checkin                           0.768813   
review_scores_communication                     1.000000   
review_scores_location                          0.558524   
review_scores_value                             0.715202   

                             review_scores_location  review_scores_value  
review_scores_rating                       0.606583             0.853106  
review_scores_accuracy                     0.594668             0.815028  
review_scores_cleanliness                  0.520925             0.722484  
review_scores_checkin                      0.541966             0.667405  
review_scores_communication                0.558524             0.715202  
review_scores_location                     1.000000             0.614159  
review_scores_value                        0.614159             1.000000

In [38]:
def review_score_with_highest_correlation_to_price_2(df_listings: pd.DataFrame, df_reviews: pd.DataFrame) -> str:
    # Define the review score columns to analyze
    review_score_columns = [
    'review_scores_rating',
    'review_scores_accuracy',
    'review_scores_cleanliness',
    'review_scores_checkin',
    'review_scores_communication',
    'review_scores_location',
    'review_scores_value',
    'price'
    ]
    df_list_takk_1 = df_listings[review_score_columns].copy()
    # df_list_takk_1['price'] = df_list_takk_1['price'].apply(lambda x: x.replace('$', '').replace(',', ''))
    # df_list_takk_1['price'] = df_list_takk_1['price'].astype(float)
    df_list_takk_1['price'] = df_list_takk_1['price'].replace({'\$': '', ',': ''}, regex=True).astype(float)
    # Calculate the correlation of the review scores with price
    correlation_with_price = df_list_takk_1[review_score_columns ].corr()['price']#+ ['price']
 

   # Find the review score with the strongest correlation to price
    strongest_correlation = correlation_with_price.drop('price').abs().idxmax()
    #strongest_correlation_value = correlation_with_price[strongest_correlation]

    #print(f"The review score with the strongest correlation to price is '{strongest_correlation}' with a correlation of {strongest_correlation_value:.2f}.")
    return strongest_correlation 

In [42]:
review_score_columns = [
'review_scores_rating',
'review_scores_accuracy',
'review_scores_cleanliness',
'review_scores_checkin',
'review_scores_communication',
'review_scores_location',
'review_scores_value',
'price'
]
df_list_takk_1 = df_listings_test[review_score_columns].copy()
df_list_takk_1

review_scores_rating  review_scores_accuracy  review_scores_cleanliness  \
0                   4.5                     4.5                        4.5   
1                   4.8                     4.8                        4.8   
2                   4.9                     4.9                        4.9   
3                   4.7                     4.7                        4.7   
4                   4.6                     4.6                        4.6   

   review_scores_checkin  review_scores_communication  review_scores_location  \
0                    4.5                          4.5                     4.5   
1                    4.8                          4.8                     4.8   
2                    4.9                          4.9                     4.9   
3                    4.7                          4.7                     4.7   
4                    4.6                          4.6                     4.6   

   review_scores_value      price  
0                  4.5    $100.00  
1                  4.8  $1,500.00  
2                  4.9    $200.00  
3                  4.7     $25.00  
4                  4.6    $182.00

In [39]:
review_score_with_highest_correlation_to_price_2(df_listings_test,df_reviews)

'review_scores_rating'

In [5]:
def prof_non_prof_host_price_diff(df_listings: pd.DataFrame, df_reviews: pd.DataFrame) -> str:

    # Load your dataset (assuming it's in a CSV file)
    # listings = pd.read_csv('path_to_your_airbnb_dataset.csv')
    # Create a new DataFrame with the required columns
    required_columns = ['host_id', 'neighbourhood_cleansed', 'price']
    listings_filtered = df_listings[required_columns].copy()

    listings_filtered['price'] = listings_filtered['price'].replace({'\$': '', ',': ''}, regex=True).astype(float)
    # Group by host and count the number of unique neighborhoods
    host_neighborhood_counts = listings_filtered.groupby('host_id')['neighbourhood_cleansed'].nunique()

    # Define professional hosts as those with more than 5 unique neighborhoods
    professional_hosts = host_neighborhood_counts[host_neighborhood_counts > 5].index

    # Create a new column to classify hosts
    listings_filtered['host_type'] = listings_filtered['host_id'].apply(lambda x: 'professional' if x in professional_hosts else 'non-professional')

    # Calculate the average price for professional and non-professional hosts
    average_prices = listings_filtered.groupby('host_type')['price'].mean().reset_index()

        # Check if there are no professional hosts
    if 'professional' not in average_prices['host_type'].values:
        return "There are no professional hosts in the dataset."

    # Check if there are no non-professional hosts
    if 'non-professional' not in average_prices['host_type'].values:
        return "There are no non-professional hosts in the dataset."


    # Calculate the price difference
    price_difference = average_prices.loc[average_prices['host_type'] == 'professional', 'price'].values[0] - \
                     average_prices.loc[average_prices['host_type'] == 'non-professional', 'price'].values[0]

    # Print the results
    #print(f"The average price difference between professional hosts and non-professional hosts is ${price_difference:.2f}.")
    return price_difference

In [79]:
prof_non_prof_host_price_diff(df_listings_test,df_reviews_test)

'There are no professional hosts in the dataset.'

In [32]:
## version2
def prof_non_prof_host_price_diff(df_listings: pd.DataFrame, df_reviews: pd.DataFrame) -> str:
    required_columns = ['host_id', 'neighbourhood_cleansed', 'price']
    listings_filtered = df_listings[required_columns].copy()

    listings_filtered['price'] = listings_filtered['price'].replace({'\$': '', ',': ''}, regex=True).astype(float)

    # Step 1: Identify professional hosts
    # Count the number of unique neighborhoods for each host
    host_neighborhood_counts = listings_filtered.groupby('host_id')['neighbourhood_cleansed'].nunique()

    # Classify hosts as professional or non-professional
    professional_hosts = host_neighborhood_counts[host_neighborhood_counts > 5].index
    listings_filtered['host_type'] = listings_filtered['host_id'].apply(lambda x: 'Professional' if x in professional_hosts else 'Non-Professional')

    # Step 2: Calculate average prices for professional and non-professional hosts
    average_prices = listings_filtered.groupby('host_type')['price'].mean()
    average_price = type(average_prices)
    # Initialize default values
    avg_professional_price = average_prices.get('Professional', 0)  # Default to 0 if not found
    avg_non_professional_price = average_prices.get('Non-Professional', 0)  # Default to 0 if not found

    # Step 3: Calculate the price difference
    #price_difference = average_prices['Professional'] - average_prices['Non-Professional']

    # Calculate the price difference
    price_difference = avg_professional_price - avg_non_professional_price

    # Output the results
    # print(f"Average price for Professional Hosts: ${average_prices['Professional']:.2f}")
    # print(f"Average price for Non-Professional Hosts: ${average_prices['Non-Professional']:.2f}")
    # print(f"Average price difference: ${price_difference:.2f}")

    return price_difference


In [33]:
prof_non_prof_host_price_diff(df_listings_test,df_reviews_test)

-401.4

In [34]:
import os

def print_directory_structure(root_dir, level=0):
    # Print the current directory
    print('    ' * level + os.path.basename(root_dir) + '/')
    
    # List all files and directories in the current directory
    try:
        for item in os.listdir(root_dir):
            item_path = os.path.join(root_dir, item)
            if os.path.isdir(item_path):
                # If it's a directory, recursively print its structure
                print_directory_structure(item_path, level + 1)
            else:
                # If it's a file, print its name
                print('    ' * (level + 1) + item)
    except PermissionError:
        print('    ' * level + '[Permission Denied]')

# Get the current working directory
current_directory = os.getcwd()

# Print the directory structure starting from the current directory
print_directory_structure(current_directory)


TURING/
    area_scrap.ipynb
    manual_test.py/
        manual_test.py
        test_df.py
        __pycache__/
            test_df.cpython-311.pyc
    sample.ipynb
    src/
        data_loader.py
        task1.py
        task2.py
        task3.py
        __init__.py
        __pycache__/
            data_loader.cpython-311.pyc
            task1.cpython-311.pyc
            task3.cpython-311.pyc
            __init__.cpython-311.pyc
    tests/
        tests_main.py
    test_data/
        test_df.py
    turing.ipynb
    __init__.py


In [115]:
import pandas as pd

def best_listing_for_revenue(listings, reviews):
    """
    Determine the listing with the best expected revenue based on the last 12 months.
    
    Parameters:
    listings (DataFrame): The Airbnb listings DataFrame containing price and minimum nights information.
    reviews (DataFrame): The Airbnb reviews DataFrame containing review counts.
    
    Returns:
    DataFrame: The listing with the best expected revenue.
    """
    review_score_columns = [
    'minimum_nights',
    'id',
    'review_scores_cleanliness',
    'review_scores_checkin',
    'review_scores_communication',
    'review_scores_location',
    'review_scores_value',
    'price'
    ]
    df_list_takk_1 = df_listings[review_score_columns].copy()
    
    # Filter listings with minimum_nights <= 7
    filtered_listings = df_list_takk_1[df_listings['minimum_nights'] <= 7].copy()

    # Calculate the number of reviews for each listing
    review_counts = df_reviews['listing_id'].value_counts().reset_index()
    review_counts.columns = ['id', 'number_of_reviews']#listing_id to id

    # Merge the review counts with the filtered listings
    merged_data = pd.merge(filtered_listings, review_counts, on='id', how='left')#listing_id to id

    # Fill NaN values in number_of_reviews with 0
    merged_data['number_of_reviews'] = merged_data['number_of_reviews'].fillna(0)

    # Calculate expected revenue
    # Assuming 60% of guests leave reviews
    merged_data['expected_revenue'] = (merged_data['number_of_reviews'] / 0.6) * merged_data['price'] * merged_data['minimum_nights']

    # Find the listing with the maximum expected revenue
    best_listing = merged_data.loc[merged_data['expected_revenue'].idxmax()]

    return filtered_listings.columns

# Example usage:
# Load your datasets (assuming they are in CSV files)
# listings = pd.read_csv('path_to_your_airbnb_listings.csv')
# reviews = pd.read_csv('path_to_your_airbnb_reviews.csv')
# best_listing = best_listing_for_revenue(listings, reviews)
# print(best_listing)


In [4]:
filtered_listings = df_listings[df_listings['minimum_nights'] <= 7].copy()
filtered_listings.columns

Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'source', 'name',
       'description', 'neighborhood_overview', 'picture_url', 'host_id',
       'host_url', 'host_name', 'host_since', 'host_location', 'host_about',
       'host_response_time', 'host_response_rate', 'host_acceptance_rate',
       'host_is_superhost', 'host_thumbnail_url', 'host_picture_url',
       'host_neighbourhood', 'host_listings_count',
       'host_total_listings_count', 'host_verifications',
       'host_has_profile_pic', 'host_identity_verified', 'neighbourhood',
       'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude',
       'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms',
       'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price',
       'minimum_nights', 'maximum_nights', 'minimum_minimum_nights',
       'maximum_minimum_nights', 'minimum_maximum_nights',
       'maximum_maximum_nights', 'minimum_nights_avg_ntm',
       'maximum_nights_avg_ntm', 'ca

In [108]:
df_reviews.head(1)

id  listing_id        date  reviewer_id reviewer_name        review  rating
0  10           1  2020-01-01          123          John  Great place!       5

In [109]:
review_counts = df_reviews['listing_id'].value_counts().reset_index()
review_counts.columns = ['id','number_of_reviews']
review_counts.head(1)

id  number_of_reviews
0   1                  1

In [110]:
merged_data = pd.merge(filtered_listings, review_counts, on='id', how='left')
merged_data.columns

Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'source', 'name',
       'description', 'neighborhood_overview', 'picture_url', 'host_id',
       'host_url', 'host_name', 'host_since', 'host_location', 'host_about',
       'host_response_time', 'host_response_rate', 'host_acceptance_rate',
       'host_is_superhost', 'host_thumbnail_url', 'host_picture_url',
       'host_neighbourhood', 'host_listings_count',
       'host_total_listings_count', 'host_verifications',
       'host_has_profile_pic', 'host_identity_verified', 'neighbourhood',
       'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude',
       'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms',
       'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price',
       'minimum_nights', 'maximum_nights', 'minimum_minimum_nights',
       'maximum_minimum_nights', 'minimum_maximum_nights',
       'maximum_maximum_nights', 'minimum_nights_avg_ntm',
       'maximum_nights_avg_ntm', 'ca

In [ ]:
merged_data['number_of_reviews'] = merged_data['number_of_reviews'].fillna(0)

In [ ]:
best_listing_for_revenue(df_listings,df_reviews)

In [39]:
review_score_with_highest_correlation_to_price_2(df_listings,df_reviews)

'review_scores_value'

In [40]:
review_score_with_highest_correlation_to_price_2(df_listings,df_reviews)

'review_scores_value'

In [64]:
prof_non_prof_host_price_diff(df_listings_test,df_reviews_test)

host_type  price
0  non-professional  401.4

## task 4


In [4]:

def median_price_premium(df_listings: pd.DataFrame, df_reviews: pd.DataFrame) -> float:
    # Filter for entire homes/apartments and other types
    entire_homes = df_listings[df_listings['room_type'] == 'Entire home/apt'].copy()
    other_listings = df_listings[df_listings['room_type'] != 'Entire home/apt'].copy()
    
    # Convert price to numeric (removing $ and commas)
    entire_homes['price'] = entire_homes['price'].replace({'\$': '', ',': ''}, regex=True).astype(float)
    other_listings['price'] = other_listings['price'].replace({'\$': '', ',': ''}, regex=True).astype(float)
    
    # Group by neighborhood and calculate median prices
    median_entire_homes = entire_homes.groupby('neighbourhood_cleansed')['price'].median()
    median_other_listings = other_listings.groupby('neighbourhood_cleansed')['price'].median()
    
    # Create a DataFrame to compare medians
    median_comparison = pd.DataFrame({
        'median_entire_homes': median_entire_homes,
        'median_other_listings': median_other_listings
    }).dropna()  # Drop neighborhoods that don't have both types of listings
    
    # Calculate the price premium
    median_comparison['price_premium'] = median_comparison['median_entire_homes'] - median_comparison['median_other_listings']
    
    # Return the average price premium across all neighborhoods
    return median_comparison['price_premium'].mean()


In [6]:
median_price_premium(df_listings,df_reviews)

86.78787878787878

In [7]:
df_listings.columns

Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'source', 'name',
       'description', 'neighborhood_overview', 'picture_url', 'host_id',
       'host_url', 'host_name', 'host_since', 'host_location', 'host_about',
       'host_response_time', 'host_response_rate', 'host_acceptance_rate',
       'host_is_superhost', 'host_thumbnail_url', 'host_picture_url',
       'host_neighbourhood', 'host_listings_count',
       'host_total_listings_count', 'host_verifications',
       'host_has_profile_pic', 'host_identity_verified', 'neighbourhood',
       'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude',
       'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms',
       'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price',
       'minimum_nights', 'maximum_nights', 'minimum_minimum_nights',
       'maximum_minimum_nights', 'minimum_maximum_nights',
       'maximum_maximum_nights', 'minimum_nights_avg_ntm',
       'maximum_nights_avg_ntm', 'ca

In [8]:
df_reviews.columns

Index(['listing_id', 'id', 'date', 'reviewer_id', 'reviewer_name', 'comments'], dtype='object')

In [13]:
df_listings.head(2)

id                         listing_url       scrape_id last_scraped  \
0  92644  https://www.airbnb.com/rooms/92644  20230906022807   2023-09-06   
1  93015  https://www.airbnb.com/rooms/93015  20230906022807   2023-09-06   

        source                                               name  \
0  city scrape  Rental unit in Earlsfield · ★4.57 · 1 bedroom ...   
1  city scrape  Rental unit in Hammersmith · ★4.82 · 2 bedroom...   

                                         description  \
0  <b>The space</b><br />Hi everyone! I have 2 ro...   
1  Gorgeous 2 bed ground floor apartment with per...   

                               neighborhood_overview  \
0                                                NaN   
1  A bit of history about the W14 area: <br />Com...   

                                         picture_url  host_id  ...  \
0  https://a0.muscache.com/pictures/715127/85bdb4...   498201  ...   
1  https://a0.muscache.com/pictures/865937ec-ee56...   499704  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                        4.67                   4.53                4.68   
1                        4.88                   4.85                4.74   

  license instant_bookable calculated_host_listings_count  \
0     NaN                t                              1   
1     NaN                f                              1   

  calculated_host_listings_count_entire_homes  \
0                                           0   
1                                           1   

  calculated_host_listings_count_private_rooms  \
0                                            1   
1                                            0   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              1.45  
1                                           0              0.27  

[2 rows x 75 columns]

## task5


In [43]:
def listings_with_best_expected_revenue(df_listings: pd.DataFrame, df_reviews: pd.DataFrame) -> int:
    review_score_columns = [
    'minimum_nights',
    'id',
    'price'
    ]
    df_list_takk_1 = df_listings[review_score_columns].copy()

    df_list_takk_1['price'] = df_list_takk_1['price'].replace({'\$': '', ',': ''}, regex=True).astype(float)
 
    # Filter listings with minimum_nights <= 7
    filtered_listings = df_list_takk_1[df_list_takk_1['minimum_nights'] <= 7]

    # Calculate the number of reviews for each listing
    review_counts = df_reviews['listing_id'].value_counts().reset_index()
    review_counts.columns = ['id', 'num_reviews']

    # Merge the review counts with the filtered listings
    merged_data = pd.merge(filtered_listings, review_counts, on='id', how='left')

    # Fill NaN values in num_reviews with 0 (for listings with no reviews)
    merged_data['num_reviews'] = merged_data['num_reviews'].fillna(0)

    # Calculate expected revenue
    # Assuming 60% of guests leave reviews, we can estimate the total number of guests
    # Total guests = num_reviews / 0.6
    # Expected revenue = Total guests * average nightly price * minimum_nights
    merged_data['expected_revenue'] = (merged_data['num_reviews'] / 0.6) * merged_data['price'] * merged_data['minimum_nights']

    # Identify the listing with the best expected revenue
    best_listing = merged_data.loc[merged_data['expected_revenue'].idxmax()]
    #result = best_listing['expected_revenue']
    result = best_listing.get('expected_revenue', 0)  # Default to 0 if not found
    # # Output the best listing details
    # print("Best Listing for Revenue:")
    # print(f"Listing ID: {best_listing['listing_id']}")
    # print(f"Expected Revenue: ${best_listing['expected_revenue']:.2f}")
    # print(f"Price per Night: ${best_listing['price']:.2f}")
    # print(f"Minimum Nights: {best_listing['minimum_nights']}")
    return int(result)


In [44]:
listings_with_best_expected_revenue(df_listings,df_reviews)

26526060

## task 6

In [12]:
def avg_diff_superhost_nonsuperhost_review_score(df_listings: pd.DataFrame, df_reviews: pd.DataFrame) -> float:
    review_score_columns = [
    'id',
    'minimum_nights',
    'host_is_superhost',
    'review_scores_rating',
    'price'
    ]
    df_list_takk_1 = df_listings[review_score_columns].copy()
    df_list_takk_1['price'] = df_list_takk_1['price'].replace({'\$': '', ',': ''}, regex=True).astype(float)

    # Rename a single column
    # df_reviews.rename(columns={'listing_id': 'id'}, inplace=True)

    # Merge listings with reviews to get review scores
    merged_data = pd.merge(df_list_takk_1, df_reviews, on='id', how='left')

    # Calculate average review scores for superhosts and normal hosts
    superhost_scores = merged_data[merged_data['host_is_superhost'] == 't']['review_scores_rating'].mean()
    normalhost_scores = merged_data[merged_data['host_is_superhost'] == 'f']['review_scores_rating'].mean()

    # Calculate the average difference
    average_difference = superhost_scores - normalhost_scores

    # Output the results
    # print("Average Review Scores:")
    # print(f"Superhosts: {superhost_scores:.2f}")
    # print(f"Normal Hosts: {normalhost_scores:.2f}")
    # print(f"Average Difference: {average_difference:.2f}")
    return average_difference



In [13]:
avg_diff_superhost_nonsuperhost_review_score(df_listings_test,df_reviews)

Average Review Scores:
Superhosts: 4.80
Normal Hosts: 4.68
Average Difference: 0.12


0.12499999999999911

## task 7

In [19]:
def host_attribute_with_second_highest_to_reviews(df_listings: pd.DataFrame, df_reviews: pd.DataFrame) -> str:

    # Select relevant columns
    columns_of_interest = ['number_of_reviews', 'host_since', 'host_listings_count', 
                        'host_identity_verified', 'calculated_host_listings_count', 
                        'host_is_superhost']

    # Filter the DataFrame to include only the relevant columns
    df = df_listings[columns_of_interest].copy()

    #df['price'] = df['price'].replace({'\\$': '', ',': ''}, regex=True).astype(float)

    # Convert categorical variables to numerical representations
    # For 'host_identity_verified' and 'host_is_superhost', we can map them to 1 and 0
    df['host_identity_verified'] = df['host_identity_verified'].map({'t': 1, 'f': 0})
    df['host_is_superhost'] = df['host_is_superhost'].map({'t': 1, 'f': 0})

    # Convert 'host_since' to a numerical value (e.g., number of years since the host started)
    df['host_since'] = pd.to_datetime(df['host_since'])
    df['host_since'] = (pd.Timestamp.now() - df['host_since']).dt.days / 365  # Convert to years

    # Calculate correlations with 'number_of_reviews'
    correlations = df.corr()['number_of_reviews'].drop('number_of_reviews')

    # Sort correlations and get the second strongest
    sorted_correlations = correlations.abs().sort_values(ascending=False)
    second_strongest = sorted_correlations.index[1]  # Get the second strongest correlation
    # Find the highest correlated attribute
    highest_correlated = correlations.idxmax()  # Get the attribute with the highest correlation
        # Output the results
    # print("Correlations with number of reviews:")
    # print(correlations)
    # print(f"\nThe second strongest correlated host attribute with number of reviews is: '{second_strongest}'")
    return highest_correlated

In [21]:
host_attribute_with_second_highest_to_reviews(df_listings_test,df_reviews_test)

'calculated_host_listings_count'